packages/dht/src/dht/routing/RoutingSession.ts

```diff
@@ -0,0 +1,222 @@
+            this.sendMoreRequests(contacts)
+        }
+    }
+
+    private sendRouteMessageRequest = async (contact: RemoteRouter): Promise<boolean> => {
+        logger.trace('sendRouteMessageRequest() sessionId: ' + this.sessionId)
+        logger.trace(`Sending routeMessage request from ${this.ownPeerDescriptor.kademliaId} to contact: ${contact.getPeerId()}`)
+        this.contactList.setContacted(contact.getPeerId())
+        this.ongoingRequests.add(contact.getPeerId().toKey())
+        if (this.mode === RoutingMode.FORWARD) {
+            return contact.forwardMessage({
+                ...this.messageToRoute,
+                previousPeer: this.ownPeerDescriptor
+            })
+        } else if (this.mode === RoutingMode.RECURSIVE_FIND) {
+            return contact.findRecursively({
+                ...this.messageToRoute,
+                previousPeer: this.ownPeerDescriptor
+            })
+        } else {
+            return contact.routeMessage({
+                ...this.messageToRoute,
+                previousPeer: this.ownPeerDescriptor
+            })
+        }
+    }
+
+    private findMoreContacts = (): RemoteRouter[] => {
+        logger.trace('findMoreContacts() sessionId: ' + this.sessionId)
+        // the contents of the connections might have changed between the rounds
+        // addContacts() will only add new contacts that were not there yet
+        const contacts = Array.from(this.connections.values())
+            .map((contact) => {
+                return new RemoteRouter(
+                    this.ownPeerDescriptor,
+                    contact.getPeerDescriptor(),
+                    toProtoRpcClient(new RoutingServiceClient(this.rpcCommunicator!.getRpcClientTransport())),
+                    contact.getServiceId()
+                )  
+            })
+        this.contactList.addContacts(contacts)
+        return this.contactList.getUncontactedContacts(this.parallelism)
+    }
+
+    public getClosestContacts = (limit: number): PeerDescriptor[] => {
+        const contacts = this.contactList.getClosestContacts(limit)
+        return contacts.map((contact) => contact.getPeerDescriptor())
+    }
+
+    private sendMoreRequests = (uncontacted: RemoteRouter[]) => {
+        logger.trace('sendMoreRequests() sessionId: ' + this.sessionId)
+        if (this.stopped) {
+            return
+        }
+        if (uncontacted.length < 1) {
+            this.emitFailure()
+            return
+        }
+        if (this.failedHopCounter >= MAX_FAILED_HOPS) {
+            logger.trace(`Stopping routing after ${MAX_FAILED_HOPS} failed attempts for sessionId: ${this.sessionId}`)
+            this.emitFailure()
+            return
+        }
+        while (this.ongoingRequests.size < this.parallelism && uncontacted.length > 0) {
+            if (this.stopped) {
+                return
+            }
+            const nextPeer = uncontacted.shift()
```

Comment:
This entity has quite many variable/class names?: peer, contact and a remote router. We could use just one of those?